subprojects/core/src/main/java/org/gradle/caching/internal/services/BuildCacheControllerFactory.java

```diff
@@ -132,27 +133,37 @@ public BuildCacheController call(BuildOperationContext context) {
         });
     }
 
-    private static BuildCacheServicesConfiguration toConfiguration(BuildCache local, BuildCacheService localService, BuildCache remote, BuildCacheService remoteService) {
-        boolean remotePush = remote != null && remote.isPush();
-        boolean localPush = local != null && local.isPush();
-        return new BuildCacheServicesConfiguration(localService, localPush, remoteService, remotePush);
+    private static BuildCacheServicesConfiguration toConfiguration(
+        @Nullable DescribedBuildCacheService<DirectoryBuildCache, DirectoryBuildCacheService> local,
+        @Nullable DescribedBuildCacheService<BuildCache, BuildCacheService> remote
+    ) {
+        boolean localPush = local != null && local.config.isPush();
+        boolean remotePush = remote != null && remote.config.isPush();
+        return new BuildCacheServicesConfiguration(
+            local != null ? local.service : null, localPush,
+            remote != null ? remote.service : null, remotePush);
     }
 
-
-    private static <T extends BuildCache> DescribedBuildCacheService createBuildCacheService(final T configuration, BuildCacheServiceRole role, Path buildIdentityPath, BuildCacheConfigurationInternal buildCacheConfiguration, Instantiator instantiator) {
-        Class<? extends BuildCacheServiceFactory<T>> castFactoryType = Cast.uncheckedCast(
+    private static <C extends BuildCache, S> DescribedBuildCacheService<C, S> createBuildCacheService(
+        C configuration,
+        BuildCacheServiceRole role,
+        Path buildIdentityPath,
+        BuildCacheConfigurationInternal buildCacheConfiguration,
+        Instantiator instantiator
+    ) {
+        Class<? extends BuildCacheServiceFactory<C>> castFactoryType = Cast.uncheckedCast(
```

Comment:
Should this be an `uncheckedNonnullCast` as well?

Reply:
Touch it you own it! I made `Instantiator` be `@NonNullApi`, too.